### PR TITLE
amp-bind: Add function to print bind's scope

### LINKED ISSUE #8719: amp-bind: Print state to console for debugging
Top-level `AMP.printState()` with a pretty-print of current state would be useful for debugging. Should handle circular references gracefully.

### PR DESCRIPTION
**Pull request by kmh287**
Adds a function enabled in dev mode to print out bind's scope.

The function allows for scope to have circular references, though currently I'm not sure that it's possible for such references to get into bind's scope. 

Fixes #8719 

/to @choumx

### CODE REVIEW

File: extensions/amp-bind/0.1/bind-impl.js [hunk 1]
```diff
@@ -910,6 +911,29 @@ export class Bind {
...
+   */
+  printAmpState_() {
+    const seen = [];
+    const seenNames = [];
+    const s = JSON.stringify(this.scope_, (key, value) => {
+      if (isObject(value)) {
+        const index = seen.indexOf(value);
+        if (index !== -1) {
+          const name = seenNames[index];
+          return `**Circular reference to '${name}'**`;
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Let's use node's handling of circular references:
```
> var a = {};
undefined
> a.a = a;
{ a: [Circular] }
```
You can remove `seenNames`.

Comment (kmh287) [inline extensions/amp-bind/0.1/bind-impl.js]: done

File: extensions/amp-bind/0.1/bind-impl.js [hunk 2]
```diff
@@ -154,6 +154,7 @@ export class Bind {
     // Expose for testing on dev.
     if (getMode().localDev) {
       AMP.reinitializeBind = this.initialize_.bind(this);
+      AMP.printAmpState = this.printAmpState_.bind(this);
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Nit: `AMP.printState`

Comment (kmh287) [inline extensions/amp-bind/0.1/bind-impl.js]: done

**Comment by zhangsu** [comment]
Sorry for chipping in here, but is this really in prod? I'm in development mode (#development=1 in the fragment) and I get this:

```
AMP.printState()
VM1646:1 Uncaught TypeError: AMP.printState is not a function
    at <anonymous>:1:5
(anonymous) @ VM1646:1
AMP.printState
undefined
AMP
Object {win: Window, config: Object, BaseElement: function, BaseTemplate: function, registerElement: function…}
```

The AMP namespace is there, but it doesn't have a `printState` attribute.

**Comment by kmh287** [comment]
@zhangsu #9460

**Comment by zhangsu** [comment]
Great, thanks! I didn't realize it was only enabled for localhost...

**Comment by stoneshower** [comment]
I get "undefined" when I do AMP.printState() on console although the page is on localhost. 
Do you know how to resolve this? Do I need to turn on some experimental modes for this?

**Comment by choumx** [comment]
@stoneshower Looks like you still need `#development=1` or `#log=4`. We should change this to use `console.info` instead.

**Comment by stoneshower** [comment]
@choumx with `#development-1` and `#log=4`, it shows the output. Thanks!

**Comment by AcidRaZor** [comment]
where is #log=4 documented? I also can't find any reference to AMP.printState on any of your documentation either. https://ampbyexample.com/components/amp-bind/

you would think if you want your technology to be adopted with the looming "google will index you higher than your competitor" axe hanging above one's head, that basic debugging tools would help.

**Comment by choumx** [comment]
@AcidRaZor Thanks for the feedback. `AMP.printState` no longer requires `#development=1` and is documented in the component docs [here](https://www.ampproject.org/docs/reference/components/amp-bind#debugging-state).

We're also working on more powerful debugging tools, e.g. #16830. I'd love to eventually have a Chrome extension similar to React Developer Tools, but unfortunately haven't had the time to implement this yet. Any interest in contributing to AMP? 😋

**Pull request merged**